modules/openxr/openxr_action_sets.cpp

```diff
@@ -0,0 +1,199 @@
+	ERR_FAIL_COND(p_action_set.is_valid());
+
+	if (action_sets.find(p_action_set) == -1) {
+		// Not sure if Vector properly refcounts so taking the long way around..
+		Ref<OpenXRActionSet> new_action_set;
+		if (action_sets.push_back(new_action_set)) {
+			action_sets.ptrw()[action_sets.size() - 1] = p_action_set;
+		}
+	}
+}
+
+void OpenXRActionSets::remove_action_set(Ref<OpenXRActionSet> p_action_set) {
+	int idx = action_sets.find(p_action_set);
+	if (idx != -1) {
+		// Not sure if Vector properly refcounts so taking the long way around..
+		action_sets.ptrw()[idx].unref();
+		action_sets.remove_at(idx);
+	}
+}
+
+void OpenXRActionSets::clear_action_sets() {
+	// Not sure if Vector properly refcounts so taking the long way around..
+	for (int i = 0; i < action_sets.size(); i++) {
+		action_sets.ptrw()[i].unref();
+	}
+	action_sets.clear();
+}
+
+void OpenXRActionSets::set_interaction_profiles(Array p_interaction_profiles) {
+	clear_interaction_profiles();
+	for (int i = 0; i < p_interaction_profiles.size(); i++) {
+		Ref<OpenXRInteractionProfile> interaction_profile = p_interaction_profiles[i];
+		add_interaction_profile(interaction_profile);
+	}
+}
+
+Array OpenXRActionSets::get_interaction_profiles() const {
+	Array arr;
+
+	for (int i = 0; i < interaction_profiles.size(); i++) {
+		arr.push_back(interaction_profiles[i]);
+	}
+
+	return arr;
+}
+
+void OpenXRActionSets::add_interaction_profile(Ref<OpenXRInteractionProfile> p_interaction_profile) {
+	ERR_FAIL_COND(p_interaction_profile.is_valid());
+
+	if (action_sets.find(p_interaction_profile) == -1) {
+		// Not sure if Vector properly refcounts so taking the long way around..
+		Ref<OpenXRInteractionProfile> new_interaction_profile;
+		if (interaction_profiles.push_back(new_interaction_profile)) {
+			interaction_profiles.ptrw()[interaction_profiles.size() - 1] = p_interaction_profile;
+		}
+	}
+}
+
+void OpenXRActionSets::remove_interaction_profile(Ref<OpenXRInteractionProfile> p_interaction_profile) {
+	int idx = interaction_profiles.find(p_interaction_profile);
+	if (idx != -1) {
+		// Not sure if Vector properly refcounts so taking the long way around..
+		interaction_profiles.ptrw()[idx].unref();
+		interaction_profiles.remove_at(idx);
+	}
+}
+
+void OpenXRActionSets::clear_interaction_profiles() {
+	// Not sure if Vector properly refcounts so taking the long way around..
+	for (int i = 0; i < interaction_profiles.size(); i++) {
+		interaction_profiles.ptrw()[i].unref();
+	}
+	interaction_profiles.clear();
+}
+
+void OpenXRActionSets::create_default_action_sets() {
+	// Create our Godot action set
+	Ref<OpenXRActionSet> action_set;
+	action_set.instantiate();
+	action_set->set_name("godot");
+	action_set->set_localised_name(String("Godot action set"));
+	add_action_set(action_set);
+
+	// Create our actions
+#define add_action(varname, actionname, actiontype, toplevel_paths) \
+	Ref<OpenXRAction>##varname;                                     \
```

Comment:
Seems to fail because of missing space?

```suggestion
	Ref<OpenXRAction> ##varname;                                    \
```

BTW we usually name our macros in uppercase.

Reply:
That space was removed by the clang-format I think... :(

Reply:
Indeed, what a mess. That should work (also moved it in scope because I didn't see a reason for it to be out?):
```diff
diff --git a/modules/openxr/openxr_action_sets.cpp b/modules/openxr/openxr_action_sets.cpp
index 9d6e23ae56..867f5547cf 100644
--- a/modules/openxr/openxr_action_sets.cpp
+++ b/modules/openxr/openxr_action_sets.cpp
@@ -152,9 +152,10 @@ void OpenXRActionSets::create_default_action_sets() {
 	add_action_set(action_set);
 
 	// Create our actions
+	// clang-format off
 #define add_action(varname, actionname, actiontype, toplevel_paths) \
-	Ref<OpenXRAction>##varname;                                     \
 	{                                                               \
+		Ref<OpenXRAction> ##varname;                                \
 		##varname.instantiate();                                    \
 		##varname->set_name(String(#varname));                      \
 		##varname->set_localised_name(String(actionname));          \
@@ -162,6 +163,7 @@ void OpenXRActionSets::create_default_action_sets() {
 		##varname->parse_toplevel_paths(String(toplevel_paths));    \
 		action_set->add_action(##varname);                          \
 	}
+	// clang-format on
 
 	add_action(trigger, "Trigger", OpenXRAction::OPENXR_ACTION_FLOAT, "/user/hand/left,/user/hand/right");
 	add_action(trigger_click, "Trigger Click", OpenXRAction::OPENXR_ACTION_BOOL, "/user/hand/left,/user/hand/right");
```

Reply:
@akien-mga I checked it in with the space but clang format is indeed complaining about it on my computer. We'll see what ci has to say about it

Reply:
See my diff again, I'm specifically disabling clang-format with `// clang-format off` (and re-enabling with `// clang-format on`).